File: apple/Sources/ReactNativeWatchOS/Modifiers/ContainerRelativeFrame.swift
import SwiftUI

/// SwiftUI `.containerRelativeFrame(_:alignment:)`. Sizes the view relative
/// to its nearest container. Requires watchOS 10+; on watchOS 9 the view is
/// returned unchanged.
enum RNWContainerRelativeFrameModifier {
    @MainActor
    static func register(into r: RNWModifierRegistry) {
        r.register("containerRelativeFrame") { view, params, _ in
            guard #available(watchOS 10.0, *) else { return view }
            let axes = axisSet(params.string("axes"))
            let alignment = RNWAlignmentParser.alignment(params.string("alignment"))
            return AnyView(view.containerRelativeFrame(axes, alignment: alignment))
        }
    }

    private static func axisSet(_ s: String?) -> Axis.Set {
        switch s {
        case "horizontal": return .horizontal
        case "vertical":   return .vertical
        default:           return [.horizontal, .vertical] // "both" / nil
        }
    }
}

File: src/modifiers/containerRelativeFrame.ts
import { createModifier } from './createModifier';

/// Axes the relative frame should expand along.
export type ContainerRelativeFrameAxes = 'horizontal' | 'vertical' | 'both';

/// Where to position the view within the container-relative frame.
export type ContainerRelativeFrameAlignment =
  | 'leading'
  | 'trailing'
  | 'center'
  | 'top'
  | 'bottom'
  | 'topLeading'
  | 'topTrailing'
  | 'bottomLeading'
  | 'bottomTrailing';

export interface ContainerRelativeFrameParams {
  /// Which axes to size relative to the nearest container. Defaults to
  /// `'both'`.
  axes?: ContainerRelativeFrameAxes;
  /// Alignment of the content inside the resulting frame. Defaults to
  /// `'center'`.
  alignment?: ContainerRelativeFrameAlignment;
}

/// SwiftUI `.containerRelativeFrame(_:alignment:)`. Sizes the view relative
/// to the nearest container (scroll view, navigation root, …). Requires
/// watchOS 10+; on older systems the view is returned unchanged.
export function containerRelativeFrame(params: ContainerRelativeFrameParams = {}) {
  return createModifier('containerRelativeFrame', params);
}
